drivers/encx24j600/encx24j600.c

```diff
@@ -0,0 +1,424 @@
+
+    return result[2] ;
+}
+#endif
+
+static void reg_set(encx24j600_t *dev, uint8_t reg, uint16_t value)
+{
+    char cmd[4] = { WCRU, reg, value, value >> 8 };
+
+//    DEBUG("encx24j600_reg_set: setting 0x%x to register 0x%x\n", (cmd[2] | (cmd[3] << 8)), reg);
+
+    enc_spi_transfer(dev, cmd, NULL, 4);
+}
+
+static void reg_set_bits(encx24j600_t *dev, uint8_t reg, uint16_t mask)
+{
+    char cmd[4] = { BFSU, reg, mask, mask >> 8 };
+
+//    DEBUG("encx24j600_reg_set: setting mask 0x%x to register 0x%x\n", (cmd[2] | (cmd[3] << 8)), reg);
+
+    enc_spi_transfer(dev, cmd, NULL, 4);
+}
+
+static void reg_clear_bits(encx24j600_t *dev, uint8_t reg, uint16_t mask)
+{
+    char cmd[4] = { BFCU, reg, mask, mask >> 8 };
+
+//    DEBUG("encx24j600_reg_set: clearing mask 0x%x to register 0x%x\n", (cmd[2] | (cmd[3] << 8)), reg);
+
+    enc_spi_transfer(dev, cmd, NULL, 4);
+}
+
+#if 0 /* unused */
+/* call with CS down (chip must already be selected) */
+static void switch_bank(encx24j600_t *dev, unsigned short bank)
+{
+    /* only switch bank if needed or possible */
+    if (dev->active_bank == bank || bank > 3) {
+        return;
+    }
+
+    spi_transfer_byte(dev->spi, B0SEL | (bank*2), NULL);
+
+    dev->active_bank = bank;
+}
+#endif
+
+/*
+ * @brief Read/Write to encx24j600's SRAM
+ *
+ * @param[in] dev   ptr to encx24j600 device handle
+ * @param[in] cmd   either WGPDATA, RGPDATA, WRXDATA, RRXDATA, WUDADATA, RUDADATA
+ * @param[in] addr  SRAM address to start reading. 0xFFFF means continue from old address
+ * @param     ptr   pointer to buffer to read from / write to
+ * @param[in] len   nr of bytes to read/write
+ */
+static void sram_op(encx24j600_t *dev, uint16_t cmd, uint16_t addr, char *ptr, int len)
+{
+    uint16_t reg;
+    char* in = NULL;
+    char* out = NULL;
+
+    /* determine pointer addr
+     *
+     * all SRAM access commands have an
+     * offset 0x5e to their pointer registers
+     * */
+    reg = cmd + 0x5e;
+
+    /* read or write? bit 1 tells us */
+    if (reg & 0x2) {
+        out = ptr;
+    } else {
+        in = ptr;
+    }
+
+    /* set pointer */
+    if (addr != 0xFFFF) {
+        reg_set(dev, reg, addr);
+    }
+
+    /* copy data */
+    cmdn(dev, cmd, in, out, len);
+}
+
+static int _init(dev_eth_t *encdev)
+{
+    encx24j600_t * dev = (encx24j600_t *) encdev;
+
+    DEBUG("encx24j600: starting initialization...\n");
+
+    /* setup IO */
+    gpio_init_out(dev->cs, GPIO_PULLUP);
+    gpio_set(dev->cs);
+    gpio_init_int(dev->int_pin, GPIO_PULLUP, GPIO_FALLING, encx24j600_isr, (void*)dev);
+
+    spi_init_master(dev->spi, SPI_CONF_FIRST_RISING, ENCX24J600_SPI_SPEED);
+
+    lock(dev);
+
+    /* initialization procedure as described in data sheet (39935c.pdf) */
+    do {
+        do {
+            vtimer_usleep(ENCX24J600_INIT_DELAY);
+            reg_set(dev, EUDAST, 0x1234);
+            vtimer_usleep(ENCX24J600_INIT_DELAY);
+        } while (reg_get(dev, EUDAST) != 0x1234);
```

Comment:
this seems dangerous to me in the current state. It would be safer if you test the SPI connection once before you go into the `do` loops - isn't there a device ID or similar fixed value register you can read out after SPI initialization to verify the connection and return an error code if the device does not respond? This is how we handle it in most other (newer) drivers...